dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h

```diff
@@ -0,0 +1,551 @@
+                    Float64 l2_reg_coef,
+                    UInt32 param_num,
+                    UInt32 batch_capacity,
+                    std::shared_ptr<IGradientComputer> gc,
+                    std::shared_ptr<IWeightsUpdater> wu)
+    : learning_rate(learning_rate),
+      l2_reg_coef(l2_reg_coef),
+      batch_capacity(batch_capacity),
+      batch_size(0),
+      gradient_computer(std::move(gc)),
+      weights_updater(std::move(wu))
+    {
+        weights.resize(param_num, Float64{0.0});
+        gradient_batch.resize(param_num + 1, Float64{0.0});
+    }
+
+    void add(const IColumn **columns, size_t row_num)
+    {
+        /// first column stores target; features start from (columns + 1)
+        const auto &target = static_cast<const ColumnVector<Float64> &>(*columns[0]).getData()[row_num];
+
+        /// Here we have columns + 1 as first column corresponds to target value, and others - to features
+        weights_updater->add_to_batch(&gradient_batch, gradient_computer,
+                                      weights, bias, learning_rate, l2_reg_coef, target, columns + 1, row_num);
+
+        ++batch_size;
+        if (batch_size == batch_capacity)
+        {
+            update_state();
+        }
+    }
+
+    void merge(const LinearModelData &rhs)
+    {
+        if (iter_num == 0 && rhs.iter_num == 0)
+            return;
+
+        update_state();
+        /// can't update rhs state because it's constant
+
+        Float64 frac = (static_cast<Float64>(iter_num) * iter_num) / (iter_num * iter_num + rhs.iter_num * rhs.iter_num);
+
+        for (size_t i = 0; i < weights.size(); ++i)
+        {
+            weights[i] = weights[i] * frac + rhs.weights[i] * (1 - frac);
+        }
+        bias = bias * frac + rhs.bias * (1 - frac);
+
+        iter_num += rhs.iter_num;
+        weights_updater->merge(*rhs.weights_updater, frac, 1 - frac);
+    }
+
+    void write(WriteBuffer &buf) const
+    {
+        writeBinary(bias, buf);
+        writeBinary(weights, buf);
+        writeBinary(iter_num, buf);
+        writeBinary(gradient_batch, buf);
+        writeBinary(batch_size, buf);
+        weights_updater->write(buf);
+    }
+
+    void read(ReadBuffer &buf)
+    {
+        readBinary(bias, buf);
+        readBinary(weights, buf);
+        readBinary(iter_num, buf);
+        readBinary(gradient_batch, buf);
+        readBinary(batch_size, buf);
+        weights_updater->read(buf);
+    }
+
+    void predict_block(ColumnVector<Float64>::Container &container, Block &block, const ColumnNumbers &arguments) const
+    {
+        gradient_computer->predict_block(container, block, arguments, weights, bias);
+    }
+
+private:
+    std::vector<Float64> weights;
+    Float64 bias{0.0};
+
+    Float64 learning_rate;
+    Float64 l2_reg_coef;
+    UInt32 batch_capacity;
+
+    UInt32 iter_num = 0;
+    std::vector<Float64> gradient_batch;
+    UInt32 batch_size;
+
+    std::shared_ptr<IGradientComputer> gradient_computer;
```

Comment:
I think, it would be better to avoid storing gradient_computer and weights_update by shared_ptr, but use GradientComputer and WeightsUpdater as part of aggregate function state (probably, inherit from them).
Thought, I am not sure we need it.